File: pine-java/src/main/java/page/liam/pine/operators/TransformByLua.java
package page.liam.pine.operators;

import page.liam.pine.*;
import page.liam.pine.metrics.Counter;
import page.liam.pine.metrics.Gauge;
import org.luaj.vm2.*;
import org.luaj.vm2.lib.*;
import org.luaj.vm2.lib.jse.JsePlatform;
import org.luaj.vm2.compiler.LuaC;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Operator: transform_by_lua
 * Metadata contract
 *   CommonInput:  [<common fields read as scalar globals>]
 *   CommonOutput: [<return values from function_for_common>]
 *   ItemInput:    [<item fields — scalars in item mode, lists in common mode>]
 *   ItemOutput:   [<return values from function_for_item>]
 */
public class TransformByLua extends AbstractOperator implements ConcurrentSafe, StatsProvider, DebugAware, MetricsAware, page.liam.pine.Closer {
    private String script;
    private String funcName;
    private boolean isItemMode;
    private LuaPool pool;
    private String operatorName;
    private boolean debug;

    /**
     * Lua compiler backend selection (system property {@code pine.lua.compiler}):
     *
     * <ul>
     *   <li>{@code luajc} (default) — LuaJ's luajc compiler
     *       ({@link org.luaj.vm2.luajc.LuaJC}): compiles Lua source directly
     *       to JVM bytecode classes, so hot scripts get the full C2 JIT
     *       treatment (inlining, loop opts). Backed by Apache BCEL.
     *   <li>{@code luac} — LuaJ's classic {@link LuaC} bytecode interpreter
     *       path. Lower one-time compile cost, slower steady-state.
     * </ul>
     *
     * <p>Scripts that luajc cannot compile (known 3.0.1 edge cases: certain
     * varargs/upvalue shapes) automatically fall back to luac — per script,
     * decided once at operator init and remembered for every pool state
     * created afterwards (see {@link LuaPool}). The fallback never changes
     * observable Lua semantics: both backends run the same LuaJ runtime,
     * libraries, and sandbox; only the execution strategy differs.
     */
    static final String COMPILER_PROP = "pine.lua.compiler";

    static boolean luajcRequested() {
        return !"luac".equalsIgnoreCase(System.getProperty(COMPILER_PROP, "luajc"));
    }

    @Override
    public void init(OperatorParams params) {
        script = params.getString("lua_script");
        String funcForItem = params.getString("function_for_item", "");
        String funcForCommon = params.getString("function_for_common", "");

        if (funcForItem.isEmpty() && funcForCommon.isEmpty()) {
            throw new IllegalArgumentException("lua: exactly one of function_for_item or function_for_common must be set");
        }
        if (!funcForItem.isEmpty() && !funcForCommon.isEmpty()) {
            throw new IllegalArgumentException("lua: cannot set both function_for_item and function_for_common");
        }

        if (!funcForItem.isEmpty()) {
            funcName = funcForItem;
            isItemMode = true;
        } else {
            funcName = funcForCommon;
            isItemMode = false;
        }

        // Validate script compiles and defines the function. This is also
        // where the luajc-vs-luac decision is made: if luajc is requested
        // (default) we try compiling + running the script under luajc once;
        // on any Error/Exception from the bytecode compiler we fall back to
        // luac for this script and remember the choice for every pool state.
        boolean useLuajc = false;
        if (luajcRequested()) {
            try {
                Globals probe = createSandboxedGlobals(true);
                probe.load(script).call();
                if (probe.get(funcName).isnil()) {
                    throw new IllegalArgumentException("lua: script does not define function \"" + funcName + "\"");
                }
                useLuajc = true;
            } catch (IllegalArgumentException e) {
                throw e;  // semantic validation failure — not a compiler issue
            } catch (Throwable t) {
                // luajc could not handle this script (BCEL generation edge
                // case). Fall through to the luac path below; if luac also
                // fails, that failure propagates as the real error.
                useLuajc = false;
            }
        }
        if (!useLuajc) {
            Globals g = createSandboxedGlobals(false);
            g.load(script).call();
            if (g.get(funcName).isnil()) {
                throw new IllegalArgumentException("lua: script does not define function \"" + funcName + "\"");
            }
        }

        pool = new LuaPool(script, useLuajc);
    }

    @Override
    public void setDebug(String operatorName, boolean debug) {
        this.operatorName = operatorName;
        this.debug = debug;
    }

    @Override
    public boolean isDebug() {
        return debug;
    }

    @Override
    public void setMetricsProvider(page.liam.pine.metrics.Provider provider) {
        String name = this.operatorName != null ? this.operatorName : "unknown";
        pool.setMetrics(
            provider.newCounter(new page.liam.pine.metrics.MetricOpts(
                "pine_lua_pool_borrow_total", "Total Lua state borrows.", "operator")).with(name),
            provider.newCounter(new page.liam.pine.metrics.MetricOpts(
                "pine_lua_pool_return_total", "Total Lua state returns.", "operator")).with(name),
            provider.newCounter(new page.liam.pine.metrics.MetricOpts(
                "pine_lua_pool_create_total", "Total Lua states created.", "operator")).with(name),
            provider.newGauge(new page.liam.pine.metrics.MetricOpts(
                "pine_lua_pool_active", "Lua states currently borrowed.", "operator")).with(name)
        );
    }

    @Override
    public void execute(CancellationToken token, OperatorInput input, OperatorOutput output) throws PineErrors.OperatorException {
        if (debug) {
            int fields = commonInput().size();
            int nonNil = 0;
            for (String f : commonInput()) {
                if (input.common(f) != null) nonNil++;
            }
            int itemCount = input.itemCount();
            String mode = isItemMode ? "item" : "common";
            logf("[pine:debug] operator=\"%s\" common_input fields=%d non_nil=%d items=%d mode=%s func=%s",
                operatorName, fields, nonNil, itemCount, mode, funcName);
        }
        Globals globals = pool.borrow();
        if (globals == null) {
            throw new PineErrors.OperatorException("lua: pool is closed");
        }
        try {
            if (isItemMode) {
                executeForItem(token, globals, input, output);
            } else {
                executeForCommon(token, globals, input, output);
            }
        } catch (LuaError e) {
            throw new PineErrors.OperatorException("lua: " + e.getMessage(), e);
        } catch (PineErrors.OperatorException e) {
            throw e;
        } catch (Exception e) {
            throw new PineErrors.OperatorException(e.getMessage(), e);
        } finally {
            pool.returnState(globals);
        }
    }

    @Override
    public Map<String, Long> operatorStats() {
        if (pool == null) return null;
        return Map.of(
                "borrow_count", pool.borrowCount.get(),
                "return_count", pool.returnCount.get(),
                "create_count", pool.createCount.get(),
                "reuse_count", pool.reuseCount.get(),
                "active_count", pool.activeCount.get()
        );
    }

    @Override
    public void close() {
        if (pool != null) {
            pool.close();
        }
    }

    private void executeForItem(CancellationToken token, Globals globals, OperatorInput input, OperatorOutput output) throws Exception {
        for (String field : commonInput()) {
            globals.set(field, toLua(input.common(field)));
        }

        LuaValue fn = globals.get(funcName);
        if (fn.isnil()) {
            throw new Exception("lua: function \"" + funcName + "\" not found");
        }

        int nret = itemOutput().size();
        int n = input.itemCount();

        // Hoist per-field columns out of the item loop: one lock + one
        // lookup per field instead of per item x field. The per-item VM
        // boundary (globals.set + invoke) is inherent to item-mode.
        List<String> fields = itemInput();
        Object[][] cols = new Object[fields.size()][];
        for (int k = 0; k < fields.size(); k++) {
            cols[k] = input.itemColumn(fields.get(k));
        }

        for (int i = 0; i < n; i++) {
            if (token.isCancelled()) break;
            for (int k = 0; k < fields.size(); k++) {
                globals.set(fields.get(k), toLua(cols[k][i]));
            }
            Varargs results;
            try {
                results = fn.invoke(LuaValue.NONE);
            } catch (LuaError e) {
                throw new Exception("lua: item[" + i + "]: " + e.getMessage(), e);
            }
            for (int j = 0; j < nret; j++) {
                Object val = fromLua(results.arg(j + 1));
                output.setItem(i, itemOutput().get(j), val);
            }
        }
    }

    private void executeForCommon(CancellationToken token, Globals globals, OperatorInput input, OperatorOutput output) throws Exception {
        if (token.isCancelled()) return;

        for (String field : commonInput()) {
            globals.set(field, toLua(input.common(field)));
        }

        int n = input.itemCount();
        for (String field : itemInput()) {
            LuaTable tbl = new LuaTable();
            Object[] col = input.itemColumn(field);
            for (int i = 0; i < n; i++) {
                tbl.set(i + 1, toLua(col[i]));
            }
            globals.set(field, tbl);
        }

        if (token.isCancelled()) return;

        LuaValue fn = globals.get(funcName);
        if (fn.isnil()) {
            throw new Exception("lua: function \"" + funcName + "\" not found");
        }

        int nret = commonOutput().size();
        Varargs results = fn.invoke(LuaValue.NONE);
        for (int j = 0; j < nret; j++) {
            Object val = fromLua(results.arg(j + 1));
            output.setCommon(commonOutput().get(j), val);
        }
    }

    private static Globals createSandboxedGlobals(boolean luajc) {
        Globals globals = new Globals();
        globals.load(new org.luaj.vm2.lib.BaseLib());
        globals.load(new org.luaj.vm2.lib.PackageLib());
        globals.load(new org.luaj.vm2.lib.TableLib());
        globals.load(new org.luaj.vm2.lib.StringLib());
        globals.load(new org.luaj.vm2.lib.MathLib());
        // LuaC always installs first: luajc's Globals.Loader still needs a
        // compiler installed for prototype parsing, and LuaJC.install only
        // swaps the loader while keeping the compiler.
        LuaC.install(globals);
        if (luajc) {
            org.luaj.vm2.luajc.LuaJC.install(globals);
        }
        globals.set("dofile", LuaValue.NIL);
        globals.set("loadfile", LuaValue.NIL);
        globals.set("require", LuaValue.NIL);
        globals.set("package", LuaValue.NIL);
        return globals;
    }

    private static LuaValue toLua(Object v) {
        if (v == null) return LuaValue.NIL;
        if (v instanceof Boolean) return LuaValue.valueOf((Boolean) v);
        if (v instanceof Number) return LuaValue.valueOf(((Number) v).doubleValue());
        if (v instanceof String) return LuaValue.valueOf((String) v);
        if (v instanceof List<?> list) {
            LuaTable tbl = new LuaTable();
            for (int i = 0; i < list.size(); i++) {
                tbl.set(i + 1, toLua(list.get(i)));
            }
            return tbl;
        }
        if (v instanceof Map<?, ?> map) {
            LuaTable tbl = new LuaTable();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                tbl.set(String.valueOf(e.getKey()), toLua(e.getValue()));
            }
            return tbl;
        }
        return LuaValue.valueOf(String.valueOf(v));
    }

    private static Object fromLua(LuaValue v) throws PineErrors.OperatorException {
        if (v.isnil()) return null;
        if (v.isboolean()) return v.toboolean();
        // Dispatch on the actual type tag, never on isnumber()/isstring():
        // luaj implements those with Lua coercion semantics, so
        // LuaString.isnumber() is true for any numeric-looking string and
        // LuaNumber.isstring() is true for every number. Routing a numeric
        // string through the number branch destroys type identity and, past
        // 2^53, the value itself (todouble round-trip) — issue #175.
        if (v.type() == LuaValue.TNUMBER) {
            double d = v.todouble();
            if (d == Math.floor(d) && !Double.isInfinite(d) && d >= Long.MIN_VALUE && d <= Long.MAX_VALUE) {
                return (long) d;
            }
            return d;
        }
        if (v.type() == LuaValue.TSTRING) return v.tojstring();
        if (v.istable()) {
            LuaTable tbl = v.checktable();
            int len = tbl.length();
            if (len > 0) {
                List<Object> arr = new ArrayList<>(len);
                for (int i = 1; i <= len; i++) {
                    arr.add(fromLua(tbl.get(i)));
                }
                return arr;
            }
            Map<String, Object> map = new LinkedHashMap<>();
            LuaValue k = LuaValue.NIL;
            while (true) {
                Varargs n = tbl.next(k);
                if ((k = n.arg1()).isnil()) break;
                if (k.type() != LuaValue.TSTRING) {
                    throw new PineErrors.OperatorException(
                            "lua: table has non-string key of type \"" + k.typename() + "\"");
                }
                map.put(k.tojstring(), fromLua(n.arg(2)));
            }
            // Lua empty table → empty array (cross-runtime convention)
            if (map.isEmpty()) return new ArrayList<>();
            return map;
        }
        return v.tojstring();
    }

    static class LuaPool {
        private final ConcurrentLinkedQueue<Globals> pool = new ConcurrentLinkedQueue<>();
        private final String initScript;
        // Compiler backend decided once at operator init (luajc with
        // verified compile, or luac fallback) — every pool state uses it.
        private final boolean luajc;
        private volatile boolean closed;
        final AtomicLong borrowCount = new AtomicLong();
        final AtomicLong returnCount = new AtomicLong();
        final AtomicLong createCount = new AtomicLong();
        final AtomicLong reuseCount = new AtomicLong();
        final AtomicLong activeCount = new AtomicLong();

        private Counter mBorrow, mReturn, mCreate;
        private Gauge mActive;

        private final Set<String> baselineKeys;
        private final ConcurrentHashMap<Globals, Map<String, LuaValue>> snapshots = new ConcurrentHashMap<>();

        LuaPool(String script, boolean luajc) {
            this.initScript = script;
            this.luajc = luajc;
            // Build baseline key set from a fresh sandboxed globals after loading script
            Globals g = createSandboxedGlobals(luajc);
            g.load(initScript).call();
            baselineKeys = snapshotKeys(g);
            pool.offer(g);
            createCount.incrementAndGet();
        }

        void setMetrics(Counter borrow, Counter ret, Counter create, Gauge active) {
            this.mBorrow = borrow;
            this.mReturn = ret;
            this.mCreate = create;
            this.mActive = active;
        }

        Globals borrow() {
            if (closed) return null;
            borrowCount.incrementAndGet();
            activeCount.incrementAndGet();
            if (mBorrow != null) mBorrow.inc();
            if (mActive != null) mActive.add(1);
            Globals g = pool.poll();
            if (g == null) {
                createCount.incrementAndGet();
                if (mCreate != null) mCreate.inc();
                g = createSandboxedGlobals(luajc);
                g.load(initScript).call();
            } else {
                reuseCount.incrementAndGet();
            }
            snapshots.put(g, snapshotBaselineValues(g));
            return g;
        }

        void returnState(Globals g) {
            returnCount.incrementAndGet();
            activeCount.decrementAndGet();
            if (mReturn != null) mReturn.inc();
            if (mActive != null) mActive.add(-1);
            if (closed) return;
            Map<String, LuaValue> snap = snapshots.remove(g);
            resetToBaseline(g, snap);
            pool.offer(g);
        }

        void close() {
            closed = true;
            pool.clear();
        }

        private Set<String> snapshotKeys(Globals g) {
            Set<String> keys = new HashSet<>();
            LuaValue k = LuaValue.NIL;
            while (true) {
                Varargs n = g.next(k);
                if ((k = n.arg1()).isnil()) break;
                if (k.isstring()) keys.add(k.tojstring());
            }
            return keys;
        }

        private Map<String, LuaValue> snapshotBaselineValues(Globals g) {
            Map<String, LuaValue> snap = new HashMap<>();
            for (String k : baselineKeys) {
                snap.put(k, g.get(k));
            }
            return snap;
        }

        private void resetToBaseline(Globals g, Map<String, LuaValue> snap) {
            if (snap == null) return;
            // Remove non-baseline keys
            Set<String> currentKeys = snapshotKeys(g);
            for (String k : currentKeys) {
                if (!baselineKeys.contains(k)) {
                    g.set(k, LuaValue.NIL);
                }
            }
            // Restore modified baseline keys
            for (Map.Entry<String, LuaValue> e : snap.entrySet()) {
                g.set(e.getKey(), e.getValue());
            }
        }
    }
}

File: pine-python/pine/operators/transform_by_lua.py
from __future__ import annotations

import sys
import threading
from collections import deque
from typing import Any

from pine.cancellation import CancellationToken
from pine.errors import OperatorException
from pine.operator import (
    AbstractOperator,
    ConcurrentSafe,
    DebugAware,
    MetricsAware,
    OperatorInput,
    OperatorOutput,
    OperatorParams,
    StatsProvider,
)

try:
    from lupa import LuaRuntime  # type: ignore[import-untyped]
    HAS_LUPA = True
except ImportError:
    HAS_LUPA = False


class TransformByLua(
    AbstractOperator, ConcurrentSafe,
    StatsProvider, DebugAware, MetricsAware,
):
    def __init__(self):
        self._script = ""
        self._func_name = ""
        self._is_item_mode = True
        self._pool: _LuaPool | None = None
        self._operator_name = ""
        self._debug = False

    def init(self, params: OperatorParams):
        if not HAS_LUPA:
            raise OperatorException("transform_by_lua: lupa package is not installed")

        self._script = params.get_string("lua_script")
        func_for_item = params.get_string("function_for_item", "")
        func_for_common = params.get_string("function_for_common", "")

        if not func_for_item and not func_for_common:
            raise ValueError(
                "lua: exactly one of function_for_item or function_for_common must be set"
            )
        if func_for_item and func_for_common:
            raise ValueError(
                "lua: cannot set both function_for_item and function_for_common"
            )

        if func_for_item:
            self._func_name = func_for_item
            self._is_item_mode = True
        else:
            self._func_name = func_for_common
            self._is_item_mode = False

        # Validate script compiles and defines the function
        runtime = _create_lua_runtime()
        runtime.execute(self._script)
        fn = runtime.globals()[self._func_name]
        if fn is None:
            raise ValueError(
                f'lua: script does not define function "{self._func_name}"'
            )

        self._pool = _LuaPool(self._script)

    def set_debug(self, name: str, enabled: bool):
        self._operator_name = name
        self._debug = enabled

    def set_metrics_provider(self, provider: Any):
        self._metrics_provider = provider

    def execute(
        self, token: CancellationToken, input_: OperatorInput,
        output: OperatorOutput,
    ) -> None:
        if self._pool is None:
            raise OperatorException("lua: pool is not initialized")

        if self._debug:
            fields = len(self.common_input())
            non_nil = sum(1 for f in self.common_input() if input_.common(f) is not None)
            item_count = input_.item_count()
            mode = "item" if self._is_item_mode else "common"
            print(
                f'[pine:debug] operator="{self._operator_name}" '
                f"common_input fields={fields} non_nil={non_nil} "
                f"items={item_count} mode={mode} func={self._func_name}",
                file=sys.stderr,
            )

        runtime = self._pool.borrow()
        if runtime is None:
            raise OperatorException("lua: pool is closed")
        try:
            if self._is_item_mode:
                self._execute_for_item(token, runtime, input_, output)
            else:
                self._execute_for_common(token, runtime, input_, output)
        except OperatorException:
            raise
        except Exception as e:
            raise OperatorException(f"lua: {e}") from e
        finally:
            self._pool.return_state(runtime)

    def operator_stats(self) -> dict[str, int]:
        if self._pool is None:
            return {}
        return {
            "borrow_count": self._pool.borrow_count,
            "return_count": self._pool.return_count,
            "create_count": self._pool.create_count,
            "active_count": self._pool.active_count,
        }

    def _execute_for_item(
        self, token: CancellationToken, runtime: LuaRuntime,
        input_: OperatorInput, output: OperatorOutput,
    ):
        g = runtime.globals()

        # Set common fields as globals
        for field in self.common_input():
            g[field] = _to_lua(runtime, input_.common(field))

        fn = g[self._func_name]
        if fn is None:
            raise OperatorException(f'lua: function "{self._func_name}" not found')

        nret = len(self.item_output())
        n = input_.item_count()

        for i in range(n):
            if token.is_cancelled():
                break
            # Set item fields as globals
            for field in self.item_input():
                g[field] = _to_lua(runtime, input_.item(i, field))

            try:
                results = fn()
            except Exception as e:
                raise OperatorException(f"lua: item[{i}]: {e}") from e

            if nret == 1:
                val = _from_lua(results)
                if val is not None:
                    output.set_item(i, self.item_output()[0], val)
            elif nret > 1:
                # Multiple returns come as a tuple
                if isinstance(results, tuple):
                    for j in range(min(nret, len(results))):
                        val = _from_lua(results[j])
                        if val is not None:
                            output.set_item(i, self.item_output()[j], val)
                else:
                    val = _from_lua(results)
                    if val is not None:
                        output.set_item(i, self.item_output()[0], val)

    def _execute_for_common(
        self, token: CancellationToken, runtime: LuaRuntime,
        input_: OperatorInput, output: OperatorOutput,
    ):
        if token.is_cancelled():
            return

        g = runtime.globals()

        # Set common fields as globals
        for field in self.common_input():
            g[field] = _to_lua(runtime, input_.common(field))

        # Set item fields as Lua tables
        n = input_.item_count()
        for field in self.item_input():
            tbl = runtime.table()
            for i in range(n):
                tbl[i + 1] = _to_lua(runtime, input_.item(i, field))
            g[field] = tbl

        if token.is_cancelled():
            return

        fn = g[self._func_name]
        if fn is None:
            raise OperatorException(f'lua: function "{self._func_name}" not found')

        nret = len(self.common_output())
        results = fn()

        if nret == 1:
            output.set_common(self.common_output()[0], _from_lua(results))
        elif nret > 1:
            if isinstance(results, tuple):
                for j in range(min(nret, len(results))):
                    output.set_common(self.common_output()[j], _from_lua(results[j]))
            else:
                output.set_common(self.common_output()[0], _from_lua(results))


def _create_lua_runtime() -> "LuaRuntime":
    """Create a sandboxed Lua runtime."""
    runtime = LuaRuntime(unpack_returned_tuples=True)
    g = runtime.globals()
    # Remove dangerous functions
    g["dofile"] = None
    g["loadfile"] = None
    g["require"] = None
    g["package"] = None
    g["io"] = None
    g["os"] = None
    return runtime


def _to_lua(runtime: "LuaRuntime", v: Any) -> Any:
    """Convert Python value to Lua-compatible value."""
    if v is None:
        return None
    if isinstance(v, bool):
        return v
    if isinstance(v, (int, float)):
        return float(v)
    if isinstance(v, str):
        return v
    if isinstance(v, (list, tuple)):
        tbl = runtime.table()
        for i, elem in enumerate(v, 1):
            tbl[i] = _to_lua(runtime, elem)
        return tbl
    if isinstance(v, dict):
        tbl = runtime.table()
        for k, val in v.items():
            tbl[str(k)] = _to_lua(runtime, val)
        return tbl
    return str(v)


def _lua_type_name(v: Any) -> str:
    """Map a lupa-returned key to its Lua type name."""
    if isinstance(v, bool):
        return "boolean"
    if isinstance(v, (int, float)):
        return "number"
    if hasattr(v, "__len__") and hasattr(v, "__getitem__"):
        return "table"
    return "userdata"


def _from_lua(v: Any) -> Any:
    """Convert Lua value to Python value."""
    if v is None:
        return None
    if isinstance(v, bool):
        return v
    if isinstance(v, float):
        d = v
        if d == int(d) and not (d == float("inf") or d == float("-inf")):
            iv = int(d)
            if -(2**63) <= iv <= 2**63 - 1:
                return iv
        return d
    if isinstance(v, int):
        return v
    if isinstance(v, str):
        return v
    # lupa table objects — check for table-like interface
    if hasattr(v, "__len__") and hasattr(v, "__getitem__"):
        try:
            length = len(v)
        except Exception:
            length = 0
        if length > 0:
            arr = []
            for i in range(1, length + 1):
                arr.append(_from_lua(v[i]))
            return arr
        # String-keyed table
        m: dict[str, Any] = {}
        for k, val in v.items():
            if not isinstance(k, str):
                lua_type_name = _lua_type_name(k)
                raise OperatorException(
                    f'lua: table has non-string key of type "{lua_type_name}"'
                )
            m[k] = _from_lua(val)
        if not m:
            # Lua empty table → empty array (cross-runtime convention)
            return []
        return m
    return str(v)


class _LuaPool:
    """Pool of Lua runtime states for reuse."""

    def __init__(self, script: str):
        self._script = script
        self._pool: deque = deque()
        self._lock = threading.Lock()
        self._closed = False
        self.borrow_count = 0
        self.return_count = 0
        self.create_count = 0
        self.active_count = 0

        # Create initial state
        rt = _create_lua_runtime()
        rt.execute(script)
        self._pool.append(rt)
        self.create_count = 1

    def borrow(self) -> "LuaRuntime | None":
        with self._lock:
            if self._closed:
                return None
            self.borrow_count += 1
            self.active_count += 1
            if self._pool:
                return self._pool.popleft()
            self.create_count += 1
        rt = _create_lua_runtime()
        rt.execute(self._script)
        with self._lock:
            if self._closed:
                self.active_count -= 1
                del rt
                return None
        return rt

    def return_state(self, runtime: "LuaRuntime"):
        with self._lock:
            self.return_count += 1
            self.active_count -= 1
            if not self._closed:
                self._pool.append(runtime)

    def close(self):
        with self._lock:
            self._closed = True
            self._pool.clear()
